controller/server/update.ml

```diff
@@ -106,149 +34,240 @@ type state =
   | ReinstallRequired
 [@@deriving sexp]
 
+type sleep_duration = float (* seconds *)
 
-(** Finite state machine handling updates *)
-let rec run ~connman ~update_url ~rauc ~set_state =
-  (* Helper to update state in signal and advance state machine *)
-  let set state =
-    set_state state; run ~connman ~update_url ~rauc ~set_state state
-  in
-  let get_proxy_uri manager = 
-      Connman.Manager.get_default_proxy manager >|= Option.map (Connman.Service.Proxy.to_uri ~include_userinfo:true)
+type config = {
+    error_backoff_duration: sleep_duration;
+    check_for_updates_interval: sleep_duration;
+}
+
+module type ServiceDeps = sig
+    module ClientI: Update_client.UpdateClientIntf
+    module RaucI: Rauc_service.RaucServiceIntf
+    val config : config
+end
+
+(** Helper to parse semver from string or fail *)
+let semver_of_string string =
+  let trimmed_string = String.trim string
   in
-  function
-  | GettingVersionInfo ->
-    (* get version information and decide what to do *)
-    let%lwt proxy = get_proxy_uri connman in
-    begin
-      match%lwt get_version_info ~proxy update_url rauc with
-      | Ok version_info ->
-
-        (* Compare latest available version to version booted. *)
-        let booted_version_compare = Semver.compare
-            (fst version_info.latest)
-            (fst version_info.booted) in
-        let booted_up_to_date = booted_version_compare = 0 in
-
-        (* Compare latest available version to version on inactive system partition. *)
-        let inactive_version_compare = Semver.compare
-            (fst version_info.latest)
-            (fst version_info.inactive) in
-        let inactive_up_to_date = inactive_version_compare = 0 in
-        let inactive_update_available = inactive_version_compare > 0 in
-
-        if booted_up_to_date || inactive_up_to_date then
-          match%lwt Rauc.get_primary rauc with
-          | Some primary_slot ->
-            if booted_up_to_date then
-              (* Don't care if inactive can be updated. I.e. Only update the inactive partition once the booted partition is outdated. This results in always two versions being available on the machine. *)
-              UpToDate version_info |> set
+  match Semver.of_string trimmed_string with
+  | None ->
+    failwith
+      (Format.sprintf "could not parse version (version string: %s)" string)
+  | Some version ->
+    version, trimmed_string
+
+module UpdateService(Deps : ServiceDeps) = struct
+    open Deps
+
+    let sleep_error_backoff =
+        Lwt_unix.sleep config.error_backoff_duration
+
+    let sleep_update_check =
+        Lwt_unix.sleep config.check_for_updates_interval
```

Comment:
Yes, since we are not running many tests or "large" quickcheck tests, it's not a big issue to have micro-sleeps. This also allows us to modify these values as needed in one place, rather than having many sleeps with hard-coded values.

In general, I'd rather not get into "mocking" concurrent / IO side-effects, because removing various blocking/IO actions can easily mask implementation bugs in tests, that might actually arise at runtime. It is also non-trivial to get it right.